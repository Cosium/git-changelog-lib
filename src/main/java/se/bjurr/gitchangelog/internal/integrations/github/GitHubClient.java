package se.bjurr.gitchangelog.internal.integrations.github;

import static com.google.common.base.Optional.absent;
import static com.google.common.base.Optional.of;
import static java.util.concurrent.TimeUnit.MINUTES;
import static org.slf4j.LoggerFactory.getLogger;

import java.util.LinkedHashMap;

import net.minidev.json.JSONArray;

import org.slf4j.Logger;

import se.bjurr.gitchangelog.internal.integrations.rest.RestClient;

import com.google.common.base.Optional;
import com.jayway.jsonpath.JsonPath;

public class GitHubClient {
 private static Logger logger = getLogger(GitHubClient.class);

 private final String api;
 private final RestClient client;

 public GitHubClient(String api) {
  this.api = api;
  this.client = new RestClient(1, MINUTES);
 }

 public Optional<GitHubIssue> getIssue(String issue) {
  try {
   if (issue.startsWith("#")) {
    issue = issue.substring(1);
   }
   String json = client.get(api + "/issues?state=all");
   JSONArray jsonArray = (JSONArray) JsonPath.read(json, "$.*");
   for (Object jsonIssue : jsonArray) {
    LinkedHashMap o = (LinkedHashMap) jsonIssue;
    String htmlUrl = o.get("html_url").toString();
    String title = o.get("title").toString();
    String number = o.get("number").toString();
    if (number.equals(issue)) {
     return of(new GitHubIssue(title, htmlUrl, number));
    }
   }
  } catch (Exception e) {
   logger.error("Error while getting issue " + issue, e);
  }
  return absent();
 }
}
