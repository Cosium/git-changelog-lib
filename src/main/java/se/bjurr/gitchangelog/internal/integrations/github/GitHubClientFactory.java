package se.bjurr.gitchangelog.internal.integrations.github;

public class GitHubClientFactory {

 private static GitHubClient gitHubClient;

 public static GitHubClient createGitHubClient(String apiUrl) {
  if (gitHubClient == null) {
   gitHubClient = new GitHubClient(apiUrl);
  }
  return gitHubClient;
 }

}
