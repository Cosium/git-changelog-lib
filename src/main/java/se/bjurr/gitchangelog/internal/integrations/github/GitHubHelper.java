package se.bjurr.gitchangelog.internal.integrations.github;

import static com.google.common.base.Optional.absent;
import static com.google.common.base.Optional.of;

import com.google.common.base.Optional;
import java.io.IOException;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import retrofit2.Call;
import retrofit2.Response;
import se.bjurr.gitchangelog.api.exceptions.GitChangelogIntegrationException;

public class GitHubHelper {

  private static Pattern PAGE_PATTERN = Pattern.compile("page=([0-9]+)>");
  private final GitHubService service;

  public GitHubHelper(GitHubService service) {
    this.service = service;
  }

  public Optional<GitHubIssue> getIssueFromAll(String issue)
      throws GitChangelogIntegrationException {
    if (issue.startsWith("#")) {
      issue = issue.substring(1);
    }

    int page = 1;
    while (page > 0) {
      Call<List<GitHubIssue>> call = service.issues(page);
      page = -1;

      try {
        Response<List<GitHubIssue>> response = call.execute();

        if (!response.isSuccessful()) {
          throw new GitChangelogIntegrationException(
              "Request:"
                  + response.raw().request().toString()
                  + "\nError:\n"
                  + response.errorBody().string());
        }

        // Pagination
        if (response.headers().get("Link") != null) {
          String link = response.headers().get("Link");
          String parsedPage = null;
          PART:
          for (String part : link.split(",")) {
            for (String piece : part.split(";")) {
              if ("rel=\"next\"".equals(piece.trim()) && parsedPage != null) {
                // Previous piece pointed to next
                page = Integer.valueOf(parsedPage);
                break PART;
              } else if (piece.contains("&page=")) {
                Matcher match = PAGE_PATTERN.matcher(piece);
                if (match.find()) {
                  parsedPage = match.group(1);
                }
              }
            }
          }
        }

        for (GitHubIssue gitHubIssue : response.body()) {
          if (issue.equals(gitHubIssue.getNumber())) {
            return of(gitHubIssue);
          }
        }

      } catch (IOException e) {
        throw new GitChangelogIntegrationException(issue, e);
      }
    }
    return absent();
  }
}
