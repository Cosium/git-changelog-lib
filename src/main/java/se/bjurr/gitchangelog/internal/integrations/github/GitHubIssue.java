package se.bjurr.gitchangelog.internal.integrations.github;

import java.util.List;

public class GitHubIssue {
  public final String title;
  public final String html_url;
  public final String number;
  private final List<GitHubLabel> labels;

  public GitHubIssue(String title, String link, String number, List<GitHubLabel> labels) {
    this.title = title;
    this.html_url = link;
    this.number = number;
    this.labels = labels;
  }

  public List<GitHubLabel> getLabels() {
    return labels;
  }

  public String getNumber() {
    return number;
  }

  public String getLink() {
    return html_url;
  }

  public String getTitle() {
    return title;
  }

  @Override
  public String toString() {
    return "GitHubIssue [title="
        + title
        + ", html_url="
        + html_url
        + ", number="
        + number
        + ", labels="
        + labels
        + "]";
  }
}
