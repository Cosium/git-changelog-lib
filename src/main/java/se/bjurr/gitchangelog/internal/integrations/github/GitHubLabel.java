package se.bjurr.gitchangelog.internal.integrations.github;

public class GitHubLabel {
  private final String name;

  public GitHubLabel(String name) {
    this.name = name;
  }

  public String getName() {
    return name;
  }

  @Override
  public String toString() {
    return "GitHubLabel [name=" + name + "]";
  }
}
