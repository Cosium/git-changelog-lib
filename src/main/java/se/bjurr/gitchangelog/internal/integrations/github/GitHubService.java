package se.bjurr.gitchangelog.internal.integrations.github;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

import java.util.List;

public interface GitHubService {
 @GET("issues?state=all&per_page=100")
 Call<List<GitHubIssue>> issues(@Query("page") int page);

 @GET("issues/{number}")
 Call<GitHubIssue> issue(@Path("number") String number);
}
