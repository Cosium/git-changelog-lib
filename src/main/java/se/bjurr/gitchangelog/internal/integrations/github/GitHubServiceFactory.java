package se.bjurr.gitchangelog.internal.integrations.github;

import static java.util.concurrent.TimeUnit.SECONDS;

import com.google.common.base.Optional;
import java.io.File;
import java.io.IOException;
import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class GitHubServiceFactory {

  private static GitHubService gitHubService = null;

  public static void reset() {
    gitHubService = null;
  }

  public static GitHubService getGitHubService(String api, final Optional<String> token) {
    return getGitHubService(api, token, null);
  }

  public static synchronized GitHubService getGitHubService(
      String api, final Optional<String> token, final Interceptor interceptor) {
    if (!api.endsWith("/")) {
      api += "/";
    }
    if (gitHubService == null) {
      File cacheDir = new File(".okhttpcache");
      cacheDir.mkdir();
      Cache cache = new Cache(cacheDir, 1024 * 1024 * 10);

      OkHttpClient.Builder builder =
          new OkHttpClient.Builder().cache(cache).connectTimeout(10, SECONDS);

      if (token != null && token.isPresent() && !token.get().isEmpty()) {
        builder.addInterceptor(
            new Interceptor() {
              @Override
              public Response intercept(Chain chain) throws IOException {
                Request original = chain.request();

                Request request =
                    original
                        .newBuilder() //
                        .addHeader("Authorization", "token " + token.get()) //
                        .method(original.method(), original.body()) //
                        .build();
                return chain.proceed(request);
              }
            });
      }

      if (interceptor != null) {
        builder.addInterceptor(interceptor);
      }

      Retrofit retrofit =
          new Retrofit.Builder() //
              .baseUrl(api) //
              .client(builder.build()) //
              .addConverterFactory(GsonConverterFactory.create()) //
              .build();

      gitHubService = retrofit.create(GitHubService.class);
    }
    return gitHubService;
  }
}
