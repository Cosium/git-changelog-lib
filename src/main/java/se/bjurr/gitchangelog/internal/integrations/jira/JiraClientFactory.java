package se.bjurr.gitchangelog.internal.integrations.jira;

public class JiraClientFactory {

 private static JiraClient jiraClient;

 public static void reset() {
  jiraClient = null;
 }

 public static JiraClient createJiraClient(String apiUrl) {
  if (jiraClient == null) {
   jiraClient = new JiraClient(apiUrl);
  }
  return jiraClient;
 }

}
