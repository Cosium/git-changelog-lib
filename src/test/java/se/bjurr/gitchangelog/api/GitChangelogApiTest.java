package se.bjurr.gitchangelog.api;

import static com.google.common.collect.ImmutableMap.of;
import static com.google.common.collect.Maps.newHashMap;
import static com.google.common.io.Resources.getResource;
import static java.nio.charset.StandardCharsets.UTF_8;
import static org.assertj.core.api.Assertions.assertThat;
import static se.bjurr.gitchangelog.api.GitChangelogApi.gitChangelogApiBuilder;
import static se.bjurr.gitchangelog.api.GitChangelogApiConstants.ZERO_COMMIT;
import static se.bjurr.gitchangelog.internal.integrations.rest.RestClient.mock;

import com.google.common.io.Resources;
import java.net.URL;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import se.bjurr.gitchangelog.internal.integrations.github.GitHubMockInterceptor;
import se.bjurr.gitchangelog.internal.integrations.github.GitHubServiceFactory;
import se.bjurr.gitchangelog.internal.integrations.jira.JiraClientFactory;
import se.bjurr.gitchangelog.internal.integrations.rest.RestClientMock;
import se.bjurr.gitchangelog.test.ApprovalsWrapper;

public class GitChangelogApiTest {
  private RestClientMock mockedRestClient;
  private GitHubMockInterceptor gitHubMockInterceptor;

  @Before
  public void before() throws Exception {
    JiraClientFactory.reset();

    this.mockedRestClient = new RestClientMock();
    this.mockedRestClient //
        .addMockedResponse(
            "/repos/tomasbjerre/git-changelog-lib/issues?state=all",
            Resources.toString(getResource("github-issues.json"), UTF_8)) //
        .addMockedResponse(
            "/jira/rest/api/2/issue/JIR-1234?fields=parent,summary,issuetype,labels,description,issuelinks",
            Resources.toString(getResource("jira-issue-jir-1234.json"), UTF_8)) //
        .addMockedResponse(
            "/jira/rest/api/2/issue/JIR-5262?fields=parent,summary,issuetype,labels,description,issuelinks",
            Resources.toString(getResource("jira-issue-jir-5262.json"), UTF_8));
    mock(this.mockedRestClient);

    this.gitHubMockInterceptor = new GitHubMockInterceptor();
    this.gitHubMockInterceptor //
        .addMockedResponse(
        "https://api.github.com/repos/tomasbjerre/git-changelog-lib/issues?state=all&per_page=100&page=1",
        Resources.toString(getResource("github-issues.json"), UTF_8));

    GitHubServiceFactory //
        .setInterceptor(this.gitHubMockInterceptor);
  }

  @After
  public void after() {
    JiraClientFactory.reset();
    GitHubServiceFactory.setInterceptor(null);
    mock(null);
  }

  @Test
  public void testThatTagsThatAreEmptyAfterCommitsHaveBeenIgnoredAreRemoved() throws Exception {
    final String templatePath = "templatetest/testAuthorsCommitsExtended.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withTemplatePath(templatePath) //
            .withIgnoreCommitsWithMessage(".*");

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testPathFilterCanBeSpecified() throws Exception {
    final String templatePath = "templatetest/testAuthorsCommitsExtended.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("1.71") //
            .withTemplatePath(templatePath) //
            .withPathFilter("src");

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testThatIssuesCanBeRemoved() throws Exception {

    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testIssuesCommits.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withSettings(settingsFile) //
            .withRemoveIssueFromMessageArgument(true) //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testThatCommitsWithoutIssueCanBeIgnoredIssuesCommits() throws Exception {

    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testIssuesCommits.mustache";

    Resources.toString(getResource(templatePath), UTF_8);

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withSettings(settingsFile) //
            .withIgnoreCommitsWithoutIssue(true) //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }

  /** The test-lightweight-2 should be ignored here. */
  @Test
  public void testThatCommitsWithoutIssueCanBeIgnoredTagsIssuesCommits() throws Exception {

    final String templatePath =
        "templatetest/testThatCommitsWithoutIssueCanBeIgnoredTagsIssuesCommits.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withGitHubIssuePattern("nooo") //
            .withGitLabIssuePattern("nooo") //
            .withCustomIssue(
                "JIRA", "JIR-[0-9]*", "http://${PATTERN_GROUP}", "${PATTERN_GROUP}") //
            .withIgnoreCommitsWithoutIssue(true) //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testThatReadableGroupMustExist() throws Exception {
    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testIssuesCommits.mustache";

    try {
      final String actual =
          gitChangelogApiBuilder() //
              .withFromCommit(ZERO_COMMIT) //
              .withSettings(settingsFile) //
              .withToRef("/test") //
              .withRemoveIssueFromMessageArgument(true) //
              .withTemplatePath(templatePath) //
              .withReadableTagName("[0-9]+?") //
              .render();
      assertThat(actual) //
          .as(
              "Should never happen! But nice to see what was rendered, if it does not crash as expected.") //
          .isEqualTo("");
    } catch (final Exception e) {
      if (!e.getMessage()
          .equals(
              "Pattern: \"[0-9]+?\" did not match any group in: \"refs/tags/test-lightweight-2\"")) {
        throw new AssertionError("Got: \"" + e.getMessage() + "\"", e);
      }
    }
  }

  @Test
  public void testThatReadableGroupCanBeSet() throws Exception {
    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testIssuesCommits.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withSettings(settingsFile) //
            .withRemoveIssueFromMessageArgument(true) //
            .withTemplatePath(templatePath) //
            .withReadableTagName(".*/([0-9]+?\\.[0-9]+?)$");

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testThatCustomVariablesCanBeUsed() throws Exception {
    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testAuthorsCommitsExtended.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withSettings(settingsFile) //
            .withExtendedVariables(newHashMap(of("customVariable", (Object) "the value"))) //
            .withRemoveIssueFromMessageArgument(true) //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void shouldUseIntegrationIfConfigured() {
    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("just title is false")) //
        .isFalse();

    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("{{title}}")) //
        .isTrue();

    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("{{link}}")) //
        .isTrue();

    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("{{type}}")) //
        .isTrue();

    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("asd{{#labels}}")) //
        .isTrue();
    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("a\nsd{{#labels}}asd\nsdasd")) //
        .isTrue();
    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("a\nsd{{labels}}asd\nsdasd")) //
        .isTrue();
  }

  @Test
  public void testThatRevertedCommitsAreRemoved() throws Exception {
    final String templatePath = "templatetest/testThatRevertedCommitsAreRemoved.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit("aa1fd33") //
            .withToCommit("4c6e078") //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }
}
