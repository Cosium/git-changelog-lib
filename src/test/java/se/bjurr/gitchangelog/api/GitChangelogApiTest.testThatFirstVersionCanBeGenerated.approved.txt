template:

# {{repoName}} changelog

Changelog of {{repoName}}.

{{#tags}}
{{#ifEquals name "Unreleased"}}{{else}}
## {{name}} ({{tagDate .}})

{{#ifContainsBreaking commits}}
### Breaking changes

{{#commits}}
{{#ifCommitBreaking .}}
- {{#eachCommitScope .}} **{{.}}** {{/eachCommitScope}} {{{commitDescription .}}} ([{{subString hash 0 5}}](https://github.com/{{ownerName}}/{{repoName}}/commit/{{hash}}) {{authorName}}) {{#eachCommitRefs .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitRefs}} {{#eachCommitFixes .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitFixes}}
{{/ifCommitBreaking}}
{{/commits}}
{{/ifContainsBreaking}}


{{#ifContainsType commits type='feat'}}
### Features

{{#commits}}
{{#ifCommitType . type='feat'}}
- {{#eachCommitScope .}} **{{.}}** {{/eachCommitScope}} {{{commitDescription .}}} ([{{subString hash 0 5}}](https://github.com/{{ownerName}}/{{repoName}}/commit/{{hash}}) {{authorName}}) {{#eachCommitRefs .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitRefs}} {{#eachCommitFixes .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitFixes}}
{{/ifCommitType}}
{{/commits}}
{{/ifContainsType}}


{{#ifContainsType commits type='fix'}}
### Bug Fixes

{{#commits}}
{{#ifCommitType . type='fix'}}
- {{#eachCommitScope .}} **{{.}}** {{/eachCommitScope}} {{{commitDescription .}}} ([{{subString hash 0 5}}](https://github.com/{{ownerName}}/{{repoName}}/commit/{{hash}}) {{authorName}}) {{#eachCommitRefs .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitRefs}} {{#eachCommitFixes .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitFixes}}
{{/ifCommitType}}
{{/commits}}
{{/ifContainsType}}


{{#ifContainsType commits type='^($|(?!fix|feat|breaking))'}}
### Other changes

{{#commits}}
{{#ifCommitType . type='^($|(?!fix|feat|breaking))'}}
**{{{messageTitle}}}**

{{#messageBodyItems}}
* {{.}} 
{{/messageBodyItems}}

[{{subString hash 0 5}}](https://github.com/{{ownerName}}/{{repoName}}/commit/{{hash}}) {{authorName}} *{{commitTime}}*

{{/ifCommitType}}
{{/commits}}
{{/ifContainsType}}

{{/ifEquals}}
{{/tags}}

---------------------------------------------

settings:

{
  "fromCommit": "0000000000000000000000000000000000000000",
  "toCommit": "0.0.1",
  "removeIssueFromMessage": false,
  "jiraEnabled": false,
  "redmineEnabled": false,
  "gitHubEnabled": false,
  "gitHubApi": "https://api.github.com/repos/tomasbjerre/git-changelog-lib",
  "extendedVariables": {},
  "ignoreCommitsWithoutIssue": false,
  "gitLabEnabled": false,
  "gitLabProjectName": "tomasbjerre",
  "useIntegrations": false
}

---------------------------------------------

changelog:

# git-changelog-lib changelog

Changelog of git-changelog-lib.



---------------------------------------------

context:

{
  "commits": [
    {
      "authorEmailAddress": "tomas.bjerre85@gmail.com",
      "authorName": "Tomas Bjerre",
      "commitTime": "2015-11-12 05:29:31",
      "commitTimeLong": 1447306171000,
      "hash": "a1aa5ff5b625e63",
      "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
      "merge": false,
      "message": "Initial commit"
    }
  ],
  "tags": [
    {
      "authors": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2015-11-12 05:29:31",
              "commitTimeLong": 1447306171000,
              "hash": "a1aa5ff5b625e63",
              "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
              "merge": false,
              "message": "Initial commit"
            }
          ],
          "authorName": "Tomas Bjerre",
          "authorEmail": "tomas.bjerre85@gmail.com"
        }
      ],
      "commits": [
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2015-11-12 05:29:31",
          "commitTimeLong": 1447306171000,
          "hash": "a1aa5ff5b625e63",
          "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
          "merge": false,
          "message": "Initial commit"
        }
      ],
      "issues": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2015-11-12 05:29:31",
              "commitTimeLong": 1447306171000,
              "hash": "a1aa5ff5b625e63",
              "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
              "merge": false,
              "message": "Initial commit"
            }
          ],
          "authors": [
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2015-11-12 05:29:31",
                  "commitTimeLong": 1447306171000,
                  "hash": "a1aa5ff5b625e63",
                  "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
                  "merge": false,
                  "message": "Initial commit"
                }
              ],
              "authorName": "Tomas Bjerre",
              "authorEmail": "tomas.bjerre85@gmail.com"
            }
          ],
          "name": "No issue",
          "title": "",
          "hasTitle": false,
          "issue": "",
          "hasIssue": false,
          "link": "",
          "hasLink": false,
          "type": "",
          "hasType": false,
          "hasDescription": false,
          "description": "",
          "hasLabels": false,
          "hasLinkedIssues": false,
          "issueType": "NOISSUE"
        }
      ],
      "issueTypes": [
        {
          "name": "No issue",
          "issues": [
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2015-11-12 05:29:31",
                  "commitTimeLong": 1447306171000,
                  "hash": "a1aa5ff5b625e63",
                  "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
                  "merge": false,
                  "message": "Initial commit"
                }
              ],
              "authors": [
                {
                  "commits": [
                    {
                      "authorEmailAddress": "tomas.bjerre85@gmail.com",
                      "authorName": "Tomas Bjerre",
                      "commitTime": "2015-11-12 05:29:31",
                      "commitTimeLong": 1447306171000,
                      "hash": "a1aa5ff5b625e63",
                      "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
                      "merge": false,
                      "message": "Initial commit"
                    }
                  ],
                  "authorName": "Tomas Bjerre",
                  "authorEmail": "tomas.bjerre85@gmail.com"
                }
              ],
              "name": "No issue",
              "title": "",
              "hasTitle": false,
              "issue": "",
              "hasIssue": false,
              "link": "",
              "hasLink": false,
              "type": "",
              "hasType": false,
              "hasDescription": false,
              "description": "",
              "hasLabels": false,
              "hasLinkedIssues": false,
              "issueType": "NOISSUE"
            }
          ],
          "type": "NOISSUE"
        }
      ],
      "name": "Unreleased",
      "tagTime": "",
      "tagTimeLong": -1,
      "hasTagTime": false
    }
  ],
  "authors": [
    {
      "commits": [
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2015-11-12 05:29:31",
          "commitTimeLong": 1447306171000,
          "hash": "a1aa5ff5b625e63",
          "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
          "merge": false,
          "message": "Initial commit"
        }
      ],
      "authorName": "Tomas Bjerre",
      "authorEmail": "tomas.bjerre85@gmail.com"
    }
  ],
  "issues": [
    {
      "commits": [
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2015-11-12 05:29:31",
          "commitTimeLong": 1447306171000,
          "hash": "a1aa5ff5b625e63",
          "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
          "merge": false,
          "message": "Initial commit"
        }
      ],
      "authors": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2015-11-12 05:29:31",
              "commitTimeLong": 1447306171000,
              "hash": "a1aa5ff5b625e63",
              "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
              "merge": false,
              "message": "Initial commit"
            }
          ],
          "authorName": "Tomas Bjerre",
          "authorEmail": "tomas.bjerre85@gmail.com"
        }
      ],
      "name": "No issue",
      "title": "",
      "hasTitle": false,
      "issue": "",
      "hasIssue": false,
      "link": "",
      "hasLink": false,
      "type": "",
      "hasType": false,
      "hasDescription": false,
      "description": "",
      "hasLabels": false,
      "hasLinkedIssues": false,
      "issueType": "NOISSUE"
    }
  ],
  "issueTypes": [
    {
      "name": "No issue",
      "issues": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2015-11-12 05:29:31",
              "commitTimeLong": 1447306171000,
              "hash": "a1aa5ff5b625e63",
              "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
              "merge": false,
              "message": "Initial commit"
            }
          ],
          "authors": [
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2015-11-12 05:29:31",
                  "commitTimeLong": 1447306171000,
                  "hash": "a1aa5ff5b625e63",
                  "hashFull": "a1aa5ff5b625e63aa5ad7b59367ec7f75658afb8",
                  "merge": false,
                  "message": "Initial commit"
                }
              ],
              "authorName": "Tomas Bjerre",
              "authorEmail": "tomas.bjerre85@gmail.com"
            }
          ],
          "name": "No issue",
          "title": "",
          "hasTitle": false,
          "issue": "",
          "hasIssue": false,
          "link": "",
          "hasLink": false,
          "type": "",
          "hasType": false,
          "hasDescription": false,
          "description": "",
          "hasLabels": false,
          "hasLinkedIssues": false,
          "issueType": "NOISSUE"
        }
      ],
      "type": "NOISSUE"
    }
  ],
  "ownerName": "tomasbjerre",
  "repoName": "git-changelog-lib"
}